src/XPurchase/VirtoCommerce.XPurchase/Extensions/RewardExtensions.cs

```diff
@@ -52,6 +52,15 @@ public static void ApplyRewards(this ShoppingCart shoppingCart, ICollection<Prom
                 shoppingCart.Discounts.Add(discount);
                 shoppingCart.DiscountAmount += discount.DiscountAmount;
             }
+
+            // remove the (added) gifts, if corresponding reward is missing
+            foreach (var lineItem in (shoppingCart.Items ?? Enumerable.Empty<LineItem>()).Where(li => li.IsGift).ToArray())
+            {
+                if (!rewards.OfType<GiftReward>().Any(re => re.ProductId == lineItem.ProductId))

```

Comment:
Add comment, please.

Reply:
yes, "// remove the (added) gifts, if corresponding reward is missing". Anything useful to add?